File: 2022/day-12/day-12.c
/* --- Advent of Code 2023 - Day 12: Hill Climbing Algorithm --- */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LENGTH 256
#define MAX_HEIGHT 100
#define MAX_WIDTH 100

/* solution part-1 */
int part_1(char input_file[]) {
    FILE *fp = fopen(input_file, "r");
    char buffer[MAX_LENGTH];
    char map[MAX_HEIGHT][MAX_WIDTH];

    /* file opened successfully */
    if (fp != NULL) {
        int start[2], end[2];
        int row, col;

        /* generate map */
        for (row = 0; fgets(buffer, MAX_LENGTH - 1, fp); row++) {
            for (col = 0; (buffer[col] >= 'a' && buffer[col] <= 'z') || (buffer[col] == 'S') || (buffer[col] == 'E'); col++) {
                map[row][col] = buffer[col];

                /* find start */
                if (buffer[col] == 'S') {
                    start[0] = col;         /* x coordinate */
                    start[1] = row;         /* y coordinate */
                    map[row][col] = 'a';
                }

                /* find end */
                if (buffer[col] == 'E') {
                    end[0] = col;           /* x coordinate */
                    end[1] = row;           /* y coordinate */
                    map[row][col] = 'z';
                }
            }
        }


        /* use dijikstra algorithm */
        int no_nodes = row * col;
        int *distance = (int *)malloc(sizeof(int) * no_nodes);      /* distance from start to node; distance[end[0] + end[1] * col] has the solution */
        int *visited = (int *)malloc(sizeof(int) * no_nodes);       /* record of visited nodes; 1 = visited, 0 = not visited */
        int **cost = (int **)malloc(sizeof(int *) * no_nodes);      /* altitude difference of nodes; 1 = reachable, ∞ = not reachable */

        for (int i = 0; i < no_nodes; i++) {
            cost[i] = (int *)malloc(sizeof(int) * no_nodes);
        }

        /* initialize distance matrix */
        for (int src = 0; src < no_nodes; src++) {
            for (int dest = 0; dest < no_nodes; dest++) {
                if (src == dest) {
                    /* distance from node to itself is 0 */
                    cost[src][dest] = 0;
                } else {
                    /* distance of all other nodes is "infinity" */
                    cost[src][dest] = no_nodes;

                    /* distance of all adjacent and by elevation recheable nodes is 1 */
                    int src_x = src % col;
                    int src_y = src / col;
                    int dest_x = dest % col;
                    int dest_y = dest / col;

                    if ((abs(src_x - dest_x) == 1 && src_y == dest_y) || (abs(src_y - dest_y) == 1 && src_x == dest_x)) {
                        if (map[dest_y][dest_x] - map[src_y][src_x] <= 1) {
                            cost[src][dest] = 1;
                        }
                    }
                }
            }
        }

        /* initialize distance and visited */
        for (int i = 0; i < no_nodes; i++) {
            distance[i] = cost[start[0] + start[1] * col][i];
            visited[i] = 0;
        }
        visited[start[0] + start[1] * col] = 1;

        /* calculate paths from start to all other nodes */
        for (int i = 0; i < no_nodes; i++) {
            int min_distance = no_nodes;
            int next_node_idx = 0;

            /* find node with minimum distance */
            for (int j = 0; j < no_nodes; j++) {
                if (visited[j] == 0 && distance[j] < min_distance) {
                    min_distance = distance[j];
                    next_node_idx = j;
                }
            }

            /* mark node as visited */
            visited[next_node_idx] = 1;

            /* update distance of all adjacent nodes */
            for (int j = 0; j < no_nodes; j++) {
                if (visited[j] == 0 && distance[j] > min_distance + cost[next_node_idx][j]) {
                    distance[j] = min_distance + cost[next_node_idx][j];
                }
            }
        }

        printf("%d\n", distance[end[0] + end[1] * col]);

        /* free memory */
        for (int i = 0; i < no_nodes; i++)
            free(cost[i]);
        free(cost);
        free(visited);
        free(distance);

    /* file not found */
    } else {
        fprintf(stderr, "Problems opening file '%s'\n", input_file);
        exit(1);
    }

    return 0;
}

/* solution part-2 */
int part_2(char input_file[]) {
    FILE *fp = fopen(input_file, "r");
    char buffer[MAX_LENGTH];
    char map[MAX_HEIGHT][MAX_WIDTH];

    /* file opened successfully */
    if (fp != NULL) {
        int start[2], end[2];
        int row, col;

        /* generate map */
        for (row = 0; fgets(buffer, MAX_LENGTH - 1, fp); row++) {
            for (col = 0; (buffer[col] >= 'a' && buffer[col] <= 'z') || (buffer[col] == 'S') || (buffer[col] == 'E'); col++) {
                map[row][col] = buffer[col];

                /* find start */
                if (buffer[col] == 'S') {
                    start[0] = col;         /* x coordinate */
                    start[1] = row;         /* y coordinate */
                    map[row][col] = 'a';
                }

                /* find end */
                if (buffer[col] == 'E') {
                    end[0] = col;           /* x coordinate */
                    end[1] = row;           /* y coordinate */
                    map[row][col] = 'z';
                }
            }
        }

        /* use dijikstra algorithm */
        int no_nodes = row * col;
        int *distance = (int *)malloc(sizeof(int) * no_nodes);      /* distance from start to node; distance[end[0] + end[1] * col] has the solution */
        int *visited = (int *)malloc(sizeof(int) * no_nodes);       /* record of visited nodes; 1 = visited, 0 = not visited */
        int **cost = (int **)malloc(sizeof(int *) * no_nodes);      /* altitude difference of nodes; 1 = reachable, ∞ = not reachable */

        for (int i = 0; i < no_nodes; i++) {
            cost[i] = (int *)malloc(sizeof(int) * no_nodes);
        }

        /* initialize distance matrix */
        for (int src = 0; src < no_nodes; src++) {
            for (int dest = 0; dest < no_nodes; dest++) {
                if (src == dest) {
                    /* distance from node to itself is 0 */
                    cost[src][dest] = 0;
                } else {
                    /* distance of all other nodes is "infinity" */
                    cost[src][dest] = no_nodes;

                    /* distance of all adjacent and by elevation recheable nodes is 1 */
                    int src_x = src % col;
                    int src_y = src / col;
                    int dest_x = dest % col;
                    int dest_y = dest / col;

                    if ((abs(src_x - dest_x) == 1 && src_y == dest_y) || (abs(src_y - dest_y) == 1 && src_x == dest_x)) {
                        if (map[dest_y][dest_x] - map[src_y][src_x] <= 1) {
                            cost[src][dest] = 1;
                        }
                    }
                }
            }
        }

        /* check all start positions for shortest path */
        int fewest_steps = no_nodes;

        for (start[0] = 0; start[0] < row; start[0]++) {
            for (start[1] = 0; start[1] < col; start[1]++) {
                if (map[start[1]][start[0]] == 'a') {

                    /* initialize distance and visited */
                    for (int i = 0; i < no_nodes; i++) {
                        distance[i] = cost[start[0] + start[1] * col][i];
                        visited[i] = 0;
                    }
                    visited[start[0] + start[1] * col] = 1;

                    /* calculate paths from start to all other nodes */
                    for (int i = 0; i < no_nodes; i++) {
                        int min_distance = no_nodes;
                        int next_node_idx = 0;

                        /* find node with minimum distance */
                        for (int j = 0; j < no_nodes; j++) {
                            if (visited[j] == 0 && distance[j] < min_distance) {
                                min_distance = distance[j];
                                next_node_idx = j;
                            }
                        }

                        /* mark node as visited */
                        visited[next_node_idx] = 1;

                        /* update distance of all adjacent nodes */
                        for (int j = 0; j < no_nodes; j++) {
                            if (visited[j] == 0 && distance[j] > min_distance + cost[next_node_idx][j]) {
                                distance[j] = min_distance + cost[next_node_idx][j];
                            }
                        }
                    }

                    /* update fewest steps */
                    fewest_steps = fewest_steps >  distance[end[0] + end[1] * col] ? distance[end[0] + end[1] * col] : fewest_steps;
                }
            }
        }

        printf("%d\n", fewest_steps);

        /* free memory */
        for (int i = 0; i < no_nodes; i++)
            free(cost[i]);
        free(cost);
        free(visited);
        free(distance);

    /* file not found */
    } else {
        fprintf(stderr, "Problems opening file '%s'\n", input_file);
        exit(1);
    }

    return 0;
}

/* main function */
int main(int argc, char *argv[]) {
    if (argc != 3) {
        fprintf(stderr, "Usage: %s -p[1|2] <input-file>\n", argv[0]);
        return 1;
    }

    if (!strcmp(argv[1], "-p1")) {
        part_1(argv[2]);
    } else if (!strcmp(argv[1], "-p2")) {
        part_2(argv[2]);
    } else {
        fprintf(stderr, "Invalid argument: '%s'\n", argv[1]);
        return 1;
    }

    return 0;
}

File: 2022/day-12/day-12.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

""" --- Advent of Code 2022 - Day 12: Hill Climbing Algorithm --- """

import sys


def get_shortest_path(start: tuple, end: tuple, map: list) -> int:
    rows = len(map)
    cols = len(map[0])
    no_nodes = rows * cols

    # use dijkstra's algorithm
    distance = [0 for _ in range(no_nodes)]
    visited = [0 for _ in range(no_nodes)]
    cost = [[0 for _ in range(no_nodes)] for _ in range(no_nodes)]

    # initialize cost matrix
    for src in range(no_nodes):
        for dest in range(no_nodes):
            if src == dest:
                # cost to get from node to itself is 0
                cost[src][dest] = 0
            else:
                # cost to get to any other node is infinity
                cost[src][dest] = no_nodes

                src_x = src % cols
                src_y = src // cols
                dest_x = dest % cols
                dest_y = dest // cols

                # cost to get to adjacent and by elevation reachable nodes is 1
                if (abs(src_x - dest_x) == 1 and src_y == dest_y) or (abs(src_y - dest_y) == 1 and src_x == dest_x):
                    if map[src_y][src_x] - map[dest_y][dest_x] <= 1:
                        cost[src][dest] = 1

    # initialize distance and visited
    for node in range(no_nodes):
        distance[node] = cost[end[0] + end[1] * cols][node]
        visited[node] = 0

    visited[end[0] + end[1] * cols] = 1

    # find shortest path
    for _ in range(no_nodes):
        min_distance = no_nodes
        next_node_idx = 0

        # find node with minimum distance
        for node in range(no_nodes):
            if (visited[node] == 0) and distance[node] < min_distance:
                min_distance = distance[node]
                next_node_idx = node

        # mark node as visited
        visited[next_node_idx] = 1

        # update distance of all adjacent nodes
        for node in range(no_nodes):
            if (visited[node] == 0) and min_distance + cost[next_node_idx][node] < distance[node]:
                distance[node] = min_distance + cost[next_node_idx][node]

    # part 1: return distance from start to end
    if start[0] != -1:
        return distance[start[0] + start[1] * cols]

    # part 2: return distance of starting point with shortest path to end
    else:
        fewest_steps = no_nodes
        for start_x in range(rows):
            for start_y in range(cols):
                if map[start_x][start_y] == ord('a'):
                    fewest_steps = min(fewest_steps, distance[start_y + start_x * cols])

        return fewest_steps


# solution part 1
def part_1():
    with open("input.txt") as f:
        input = f.read().splitlines()

    map = [[0 for _ in range(len(input[0]))] for _ in range(len(input))]

    # get start and end points
    for row in range(len(input)):
        for col in range(len(input[0])):
            map[row][col] = ord(input[row][col])

            if input[row][col] == "S":
                start = (col, row)
                map[row][col] = ord('a')
            if input[row][col] == "E":
                end = (col, row)
                map[row][col] = ord('z')

    print(get_shortest_path(start, end, map))


# solution part 2
def part_2():
    with open("input.txt") as f:
        input = f.read().splitlines()

    map = [[0 for _ in range(len(input[0]))] for _ in range(len(input))]

    # get start and end points
    for row in range(len(input)):
        for col in range(len(input[0])):
            map[row][col] = ord(input[row][col])

            if input[row][col] == "S":
                start = (-1, -1)
                map[row][col] = ord('a')
            if input[row][col] == "E":
                end = (col, row)
                map[row][col] = ord('z')

    # find starting point with shortest path
    print(get_shortest_path(start, end, map))


# main entry point
if len(sys.argv) > 1 and sys.argv[1] == "-p1":
    part_1()
elif len(sys.argv) > 1 and sys.argv[1] == "-p2":
    part_2()
else:
    print("Usage: python", sys.argv[0], "[-p1|-p2]")
    sys.exit(1)
